File: src/tables/glyf.cpp
#include <bitset>

#include <QVarLengthArray>

#include "src/algo.h"

#include "tables.h"

struct NegativeUInt8
{
    static const int Size = 1;
    static const QString Type;

    static NegativeUInt8 parse(const uint8_t *data)
    { return { data[0] }; }

    static QString toString(const NegativeUInt8 &value)
    { return numberToString(-qint16(value.d)); }

    DEFAULT_DEBUG(NegativeUInt8)

    operator uint8_t() const { return d; }

    uint8_t d;
};

const QString NegativeUInt8::Type = UInt8::Type;


struct SimpleGlyphFlags
{
    static const int Size = 1;
    static const QString Type;

    enum Flag : quint8
    {
        ON_CURVE_POINT = 0x01,
        X_SHORT_VECTOR = 0x02,
        Y_SHORT_VECTOR = 0x04,
        REPEAT_FLAG = 0x08,
        X_IS_SAME_OR_POSITIVE_X_SHORT_VECTOR = 0x10,
        Y_IS_SAME_OR_POSITIVE_Y_SHORT_VECTOR = 0x20,
        OVERLAP_SIMPLE = 0x40,
    };

    Q_DECLARE_FLAGS(Flags, Flag)

    static SimpleGlyphFlags parse(const quint8 *data)
    {
        return { Flag(data[0]) };
    }

    static QString toString(const SimpleGlyphFlags &value)
    {
        // This method would be called a lot, therefore we have to use a cache.
        //
        // Valid flags would be in a 0..64 range, so using a QVarLengthArray
        // is more performant than QHash.
        static QVarLengthArray<QString, 64> cache;

        if (cache.isEmpty()) {
            cache.resize(64);
        }

        if (value.d < cache.size() && !cache[value.d].isEmpty()) {
            return cache[value.d];
        }

        std::bitset<8> bits(value.d);
        auto flagsStr = QString::fromUtf8(bits.to_string().c_str()) + '\n';

        if (bits[0]) flagsStr += "Bit 0: On curve point\n";
        if (bits[1]) flagsStr += "Bit 1: X-coordinate is 1 byte long\n";
        if (bits[2]) flagsStr += "Bit 2: Y-coordinate is 1 byte long\n";
        if (bits[3]) flagsStr += "Bit 3: Repeat flag\n";

        if (bits[1] && bits[4]) flagsStr += "Bit 4: X-coordinate is positive\n";
        if (bits[1] && !bits[4]) flagsStr += "Bit 4: X-coordinate is negative\n";
        if (!bits[1] && bits[4]) flagsStr += "Bit 4: Use the previous X-coordinate\n";
        if (!bits[1] && !bits[4]) flagsStr += "Bit 4: X-coordinate is 2 byte long, signed\n";

        if (bits[2] && bits[5]) flagsStr += "Bit 5: Y-coordinate is positive\n";
        if (bits[2] && !bits[5]) flagsStr += "Bit 5: Y-coordinate is negative\n";
        if (!bits[2] && bits[5]) flagsStr += "Bit 5: Use the previous Y-coordinate\n";
        if (!bits[2] && !bits[5]) flagsStr += "Bit 5: Y-coordinate is 2 byte long, signed\n";

        if (bits[6]) flagsStr += "Bit 6: Contours may overlap\n";
        // 7 - reserved

        flagsStr.chop(1); // trim trailing newline

        if (value.d < cache.size()) {
            cache[value.d] = flagsStr;
        }

        return flagsStr;
    }

    operator Flag() const { return d; }

    Flag d;
};

const QString SimpleGlyphFlags::Type = Parser::BitflagsType;


struct CompositeGlyphFlags
{
    static const int Size = 2;
    static const QString Type;

    enum Flag : quint16
    {
        ARG_1_AND_2_ARE_WORDS = 0x0001,
        ARGS_ARE_XY_VALUES = 0x0002,
        ROUND_XY_TO_GRID = 0x0004,
        WE_HAVE_A_SCALE = 0x0008,
        MORE_COMPONENTS = 0x0020,
        WE_HAVE_AN_X_AND_Y_SCALE = 0x0040,
        WE_HAVE_A_TWO_BY_TWO = 0x0080,
        WE_HAVE_INSTRUCTIONS = 0x0100,
        USE_MY_METRICS = 0x0200,
        OVERLAP_COMPOUND = 0x0400,
        SCALED_COMPONENT_OFFSET = 0x0800,
        UNSCALED_COMPONENT_OFFSET = 0x1000,
   };

    Q_DECLARE_FLAGS(Flags, Flag)

    static CompositeGlyphFlags parse(const quint8 *data)
    {
        return { Flag(quint16(UInt16::parse(data))) };
    }

    static QString toString(const CompositeGlyphFlags &value)
    {
        std::bitset<16> bits(value.d);
        auto flagsStr = QString::fromUtf8(bits.to_string().c_str()) + '\n';

        if (bits[0]) flagsStr += "Bit 0: Arguments are 16-bit\n";
        if (bits[1]) flagsStr += "Bit 1: Arguments are signed xy values\n";
        if (bits[2]) flagsStr += "Bit 2: Round XY to grid\n";
        if (bits[3]) flagsStr += "Bit 3: Has a simple scale\n";
        // 4 - reserved
        if (bits[5]) flagsStr += "Bit 5: Has more glyphs\n";
        if (bits[6]) flagsStr += "Bit 6: Non-propotional scale\n";
        if (bits[7]) flagsStr += "Bit 7: Has 2 by 2 transformation matrix\n";
        if (bits[8]) flagsStr += "Bit 8: Has instructions after the last component\n";
        if (bits[9]) flagsStr += "Bit 9: Use my metrics\n";
        if (bits[10]) flagsStr += "Bit 10: Components overlap\n";
        if (bits[11]) flagsStr += "Bit 11: Scaled component offset\n";
        if (bits[12]) flagsStr += "Bit 12: Unscaled component offset\n";
        // 13, 14, 15 - reserved

        flagsStr.chop(1); // trim trailing newline

        return flagsStr;
    }

    operator Flag() const { return d; }

    Flag d;
};

const QString CompositeGlyphFlags::Type = Parser::BitflagsType;


static void parseSimpleGlyph(const quint16 numOfContours, Parser &parser)
{
    quint16 lastPoint = 0;
    parser.readArray("Endpoints", numOfContours, [&](const auto index){
        lastPoint = parser.read<UInt16>(index);
    });

    const auto instructionLength = parser.read<UInt16>("Instructions size");
    parser.readBytes("Instructions", instructionLength);

    parser.beginGroup("Flags");
    QVarLengthArray<SimpleGlyphFlags> allFlags;
    const auto totalPoints = lastPoint + 1;
    auto pointsLeft = totalPoints;
    while (pointsLeft > 0) {
        const auto flags = parser.read<SimpleGlyphFlags>("Flag");
        allFlags << flags;

        auto repeats = 1;
        if (flags & SimpleGlyphFlags::REPEAT_FLAG) {
            repeats = parser.read<UInt8>("Number of repeats");
            for (int i = 0; i < repeats; ++i) {
                allFlags << flags;
            }

            repeats += 1;
        }

        pointsLeft -= repeats;
    }
    parser.endGroup();

    {
        // The number of coordinates can be lower than the number of flags.
        int xCoords = 0;
        for (const auto flags : allFlags) {
            if (flags & SimpleGlyphFlags::X_SHORT_VECTOR) {
                if (flags & SimpleGlyphFlags::X_IS_SAME_OR_POSITIVE_X_SHORT_VECTOR) {
                    xCoords += 1;
                } else {
                    xCoords += 1;
                }
            } else {
                if (flags & SimpleGlyphFlags::X_IS_SAME_OR_POSITIVE_X_SHORT_VECTOR) {
                    // Nothing.
                } else {
                    xCoords += 1;
                }
            }
        }

        parser.beginArray("X-coordinates", xCoords);
        for (auto [index, flags] : algo::enumerate(allFlags)) {
            if (*flags & SimpleGlyphFlags::X_SHORT_VECTOR) {
                if (*flags & SimpleGlyphFlags::X_IS_SAME_OR_POSITIVE_X_SHORT_VECTOR) {
                    parser.read<UInt8>(index);
                } else {
                    parser.read<NegativeUInt8>(index);
                }
            } else {
                if (*flags & SimpleGlyphFlags::X_IS_SAME_OR_POSITIVE_X_SHORT_VECTOR) {
                    // Nothing.
                } else {
                    parser.read<Int16>(index);
                }
            }
        }
        parser.endArray();
    }

    {
        // The number of coordinates can be lower than the number of flags.
        int yCoords = 0;
        for (const auto flags : allFlags) {
            if (flags & SimpleGlyphFlags::Y_SHORT_VECTOR) {
                if (flags & SimpleGlyphFlags::Y_IS_SAME_OR_POSITIVE_Y_SHORT_VECTOR) {
                    yCoords += 1;
                } else {
                    yCoords += 1;
                }
            } else {
                if (flags & SimpleGlyphFlags::Y_IS_SAME_OR_POSITIVE_Y_SHORT_VECTOR) {
                    // Nothing.
                } else {
                    yCoords += 1;
                }
            }
        }

        parser.beginArray("Y-coordinates", yCoords);
        for (auto [index, flags] : algo::enumerate(allFlags)) {
            if (*flags & SimpleGlyphFlags::Y_SHORT_VECTOR) {
                if (*flags & SimpleGlyphFlags::Y_IS_SAME_OR_POSITIVE_Y_SHORT_VECTOR) {
                    parser.read<UInt8>(index);
                } else {
                    parser.read<NegativeUInt8>(index);
                }
            } else {
                if (*flags & SimpleGlyphFlags::Y_IS_SAME_OR_POSITIVE_Y_SHORT_VECTOR) {
                    // Nothing.
                } else {
                    parser.read<Int16>(index);
                }
            }
        }
        parser.endArray();
    }
}

static void parseCompositeGlyph(Parser &parser)
{
    const auto flags = parser.read<CompositeGlyphFlags>("Flag");
    parser.read<GlyphId>("Glyph ID");

    std::array<double, 6> matrix = {0};
    bool hasTs = false;

    if (flags & CompositeGlyphFlags::ARGS_ARE_XY_VALUES) {
        parser.beginGroup();
        hasTs = true;
        if (flags & CompositeGlyphFlags::ARG_1_AND_2_ARE_WORDS) {
            matrix[4] = parser.read<Int16>("E");
            matrix[5] = parser.read<Int16>("F");
        } else {
            matrix[4] = parser.read<Int8>("E");
            matrix[5] = parser.read<Int8>("F");
        }
    } else {
        if (flags & CompositeGlyphFlags::ARG_1_AND_2_ARE_WORDS) {
            parser.read<UInt16>("Point 1");
            parser.read<UInt16>("Point 2");
        } else {
            parser.read<UInt8>("Point 1");
            parser.read<UInt8>("Point 2");
        }
    }

    if (flags & CompositeGlyphFlags::WE_HAVE_A_TWO_BY_TWO) {
        if (!hasTs) {
            parser.beginGroup();
        }
        hasTs = true;

        matrix[0] = parser.read<F2DOT14>("A");
        matrix[1] = parser.read<F2DOT14>("B");
        matrix[2] = parser.read<F2DOT14>("C");
        matrix[3] = parser.read<F2DOT14>("D");
    } else if (flags & CompositeGlyphFlags::WE_HAVE_AN_X_AND_Y_SCALE) {
        if (!hasTs) {
            parser.beginGroup();
        }
        hasTs = true;

        matrix[0] = parser.read<F2DOT14>("A");
        matrix[3] = parser.read<F2DOT14>("D");
    } else if (flags & CompositeGlyphFlags::WE_HAVE_A_SCALE) {
        if (!hasTs) {
            parser.beginGroup();
        }
        hasTs = true;

        matrix[0] = parser.read<F2DOT14>("A");
        matrix[3] = matrix[0];
    }

    if (hasTs) {
        const auto matrixStr = QString("%1 %2 %3 %4 %5 %6")
            .arg(floatToString(matrix.at(0)))
            .arg(floatToString(matrix.at(1)))
            .arg(floatToString(matrix.at(2)))
            .arg(floatToString(matrix.at(3)))
            .arg(floatToString(matrix.at(4)))
            .arg(floatToString(matrix.at(5)));
        parser.endGroup("Matrix", matrixStr);
    }

    if (flags & CompositeGlyphFlags::MORE_COMPONENTS) {
        parseCompositeGlyph(parser);
    } else if (flags & CompositeGlyphFlags::WE_HAVE_INSTRUCTIONS) {
         const auto size = parser.read<UInt16>("Number of instructions");
         parser.readBytes("Instructions", size);
    }
}

void parseGlyf(const quint16 numberOfGlyphs, const QVector<quint32> &glyphOffsets, Parser &parser)
{
    Q_ASSERT(int(numberOfGlyphs) + 1 == glyphOffsets.size());

    const auto tableStart = parser.offset();

    // Glyphs can be empty, therefore the real number of glyphs can be lower than numberOfGlyphs.
    int glyphsCount = 0;
    for (quint16 i = 0; i < numberOfGlyphs; i++) {
        if (glyphOffsets[i] != glyphOffsets[i + 1]) {
            glyphsCount += 1;
        }
    }

    parser.beginArray("Glyphs", glyphsCount);
    for (quint16 index = 0; index < numberOfGlyphs; index++) {
        const auto start = tableStart + glyphOffsets[index];
        const auto end = tableStart + glyphOffsets[index + 1];
        if (start == end) {
            continue;
        }

        parser.beginGroup();

        const auto numberOfContours = parser.read<Int16>("Number of contours");
        parser.read<Int16>("x min");
        parser.read<Int16>("y min");
        parser.read<Int16>("x max");
        parser.read<Int16>("y max");

        if (numberOfContours == 0) {
            // Empty.
        } else if (numberOfContours > 0) {
            parseSimpleGlyph(quint16(numberOfContours), parser);
        } else {
            parseCompositeGlyph(parser);
        }

        if (parser.offset() < end) {
            const auto diff = end - parser.offset();
            if (diff < 4) {
                parser.readPadding(diff);
            } else {
                parser.readUnsupported(diff);
            }
        }

        if (numberOfContours == 0) {
            parser.endGroup(QString("Glyph %1 (empty)").arg(index));
        } else if (numberOfContours > 0) {
            parser.endGroup(QString("Glyph %1").arg(index));
        } else {
            parser.endGroup(QString("Glyph %1 (composite)").arg(index));
        }
    }
    parser.endArray();
}

File: core/src/tables/glyf.rs
use std::ops::Range;

use crate::parser::*;
use crate::Result;

#[derive(Clone, Copy, PartialEq, Debug)]
enum GlyphType {
    Empty,
    Simple,
    Composite,
}


// A special case, where `u8` should be negative but not i8.
#[derive(Clone, Copy, Debug)]
struct NegativeU8(i16);

impl FromData for NegativeU8 {
    const SIZE: usize = 1;
    const NAME: &'static str = "UInt8";

    #[inline]
    fn parse(data: &[u8]) -> Result<Self> {
        Ok(NegativeU8(-(data[0] as i16)))
    }
}

impl std::fmt::Display for NegativeU8 {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)
    }
}


#[derive(Clone, Copy, Debug)]
struct SimpleGlyphFlags(u8);

impl SimpleGlyphFlags {
    #[inline] fn x_short(self) -> bool { self.0 & 0x02 != 0 }
    #[inline] fn y_short(self) -> bool { self.0 & 0x04 != 0 }
    #[inline] fn repeat_flag(self) -> bool { self.0 & 0x08 != 0 }
    #[inline] fn x_is_same_or_positive_short(self) -> bool { self.0 & 0x10 != 0 }
    #[inline] fn y_is_same_or_positive_short(self) -> bool { self.0 & 0x20 != 0 }
}

impl FromData for SimpleGlyphFlags {
    const NAME: &'static str = "BitFlags";

    #[inline]
    fn parse(data: &[u8]) -> Result<Self> {
        u8::parse(data).map(SimpleGlyphFlags)
    }
}

impl std::fmt::Display for SimpleGlyphFlags {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use std::fmt::Write;

        let mut s = String::new();

        let bits = U8Bits(self.0);
        writeln!(&mut s, "{}", bits)?;
        if bits[0] { s.push_str("Bit 0: On curve point\n"); }
        if bits[1] { s.push_str("Bit 1: X-coordinate is 1 byte long\n"); }
        if bits[2] { s.push_str("Bit 2: Y-coordinate is 1 byte long\n"); }
        if bits[3] { s.push_str("Bit 3: Repeat flag\n"); }

        if bits[1] && bits[4] { s.push_str("Bit 4: X-coordinate is positive\n"); }
        if bits[1] && !bits[4] { s.push_str("Bit 4: X-coordinate is negative\n"); }
        if !bits[1] && bits[4] { s.push_str("Bit 4: Use the previous X-coordinate\n"); }
        if !bits[1] && !bits[4] { s.push_str("Bit 4: X-coordinate is 2 byte long, signed\n"); }

        if bits[2] && bits[5] { s.push_str("Bit 5: Y-coordinate is positive\n"); }
        if bits[2] && !bits[5] { s.push_str("Bit 5: Y-coordinate is negative\n"); }
        if !bits[2] && bits[5] { s.push_str("Bit 5: Use the previous Y-coordinate\n"); }
        if !bits[2] && !bits[5] { s.push_str("Bit 5: Y-coordinate is 2 byte long, signed\n"); }

        if bits[6] { s.push_str("Bit 6: Contours may overlap\n"); }
        // 7 - reserved

        s.pop(); // pop last new line

        f.write_str(&s)
    }
}


#[derive(Clone, Copy, Debug)]
struct CompositeGlyphFlags(u16);

impl CompositeGlyphFlags {
    #[inline] fn arg_1_and_2_are_words(self) -> bool { self.0 & 0x0001 != 0 }
    #[inline] fn args_are_xy_values(self) -> bool { self.0 & 0x0002 != 0 }
    #[inline] fn we_have_a_scale(self) -> bool { self.0 & 0x0008 != 0 }
    #[inline] fn more_components(self) -> bool { self.0 & 0x0020 != 0 }
    #[inline] fn we_have_an_x_and_y_scale(self) -> bool { self.0 & 0x0040 != 0 }
    #[inline] fn we_have_a_two_by_two(self) -> bool { self.0 & 0x0080 != 0 }
    #[inline] fn we_have_instructions(self) -> bool { self.0 & 0x0100 != 0 }
}

impl FromData for CompositeGlyphFlags {
    const NAME: &'static str = "BitFlags";

    #[inline]
    fn parse(data: &[u8]) -> Result<Self> {
        u16::parse(data).map(CompositeGlyphFlags)
    }
}

impl std::fmt::Display for CompositeGlyphFlags {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use std::fmt::Write;
        let mut s = String::new();

        let bits = U16Bits(self.0);
        writeln!(&mut s, "{}", bits)?;
        if bits[0]  { s.push_str("Bit 0: Arguments are 16-bit\n"); }
        if bits[1]  { s.push_str("Bit 1: Arguments are signed xy values\n"); }
        if bits[2]  { s.push_str("Bit 2: Round XY to grid\n"); }
        if bits[3]  { s.push_str("Bit 3: Has a simple scale\n"); }
        // 4 - reserved
        if bits[5]  { s.push_str("Bit 5: Has more glyphs\n"); }
        if bits[6]  { s.push_str("Bit 6: Non-proportional scale\n"); }
        if bits[7]  { s.push_str("Bit 7: Has 2 by 2 transformation matrix\n"); }
        if bits[8]  { s.push_str("Bit 8: Has instructions after the last component\n"); }
        if bits[9]  { s.push_str("Bit 9: Use my metrics\n"); }
        if bits[10] { s.push_str("Bit 10: Components overlap\n"); }
        if bits[11] { s.push_str("Bit 11: Scaled component offset\n"); }
        if bits[12] { s.push_str("Bit 12: Unscaled component offset\n"); }
        // 13, 14, 15 - reserved

        s.pop(); // pop last new line

        f.write_str(&s)
    }
}


pub fn parse(ranges: &[Range<usize>], parser: &mut Parser) -> Result<()> {
    let start = parser.offset();

    for (i, range) in ranges.iter().enumerate() {
        // Skip empty ranges here and not in loca::collect_ranges,
        // so we can maintain a proper glyph index.
        if range.len() == 0 {
            continue;
        }

        parser.jump_to(start + range.start)?;
        parser.begin_group("");
        let kind = parse_glyph(parser)?;
        if kind == GlyphType::Composite {
            parser.end_group_with_title(format!("Glyph {} (composite)", i));
        } else {
            parser.end_group_with_title(format!("Glyph {}", i));
        }
    }

    Ok(())
}

fn parse_glyph(parser: &mut Parser) -> Result<GlyphType> {
    let num_of_contours = parser.read::<i16>("Number of contours")?;
    parser.read::<i16>("x min")?;
    parser.read::<i16>("y min")?;
    parser.read::<i16>("x max")?;
    parser.read::<i16>("y max")?;

    if num_of_contours == 0 {
        Ok(GlyphType::Empty)
    } else if num_of_contours > 0 {
        parse_simple_glyph(num_of_contours as u16, parser)?;
        Ok(GlyphType::Simple)
    } else {
        parse_composite_glyph(parser)?;
        Ok(GlyphType::Composite)
    }
}

fn parse_simple_glyph(num_of_contours: u16, parser: &mut Parser) -> Result<()> {
    let mut last_point = 0;
    parser.begin_group("Endpoints");
    for i in 0..num_of_contours {
        last_point = parser.read2::<u16>(format!("Endpoint {}", i))?;
    }
    parser.end_group();

    let instructions_len = parser.read::<u16>("Instructions size")?;
    if instructions_len > 0 {
        parser.read_bytes(instructions_len as usize, "Instructions")?;
    }

    parser.begin_group("Flags");
    let mut all_flags = Vec::new();
    let mut has_x_coords = false;
    let mut has_y_coords = false;
    let total_points = last_point + 1;
    let mut points_left = total_points;
    while points_left > 0 {
        let flags = parser.read::<SimpleGlyphFlags>("Flag")?;
        all_flags.push(flags);

        if flags.x_short() || !flags.x_is_same_or_positive_short() {
            has_x_coords = true;
        }

        if flags.y_short() || !flags.y_is_same_or_positive_short() {
            has_y_coords = true;
        }

        let repeats = if flags.repeat_flag() {
            let repeats = parser.read::<u8>("Number of repeats")?;
            for _ in 0..repeats {
                all_flags.push(flags);
            }

            repeats as u16 + 1
        } else {
            1
        };

        points_left -= repeats;
    }
    parser.end_group();

    if has_x_coords {
        parser.begin_group("X-coordinates");
        for flags in &all_flags {
            if flags.x_short() {
                if flags.x_is_same_or_positive_short() {
                    parser.read::<u8>("Coordinate")?;
                } else {
                    parser.read::<NegativeU8>("Coordinate")?;
                }
            } else {
                if flags.x_is_same_or_positive_short() {
                    // Nothing.
                } else {
                    parser.read::<i16>("Coordinate")?;
                }
            }
        }
        parser.end_group();
    }

    if has_y_coords {
        parser.begin_group("Y-coordinates");
        for flags in &all_flags {
            if flags.y_short() {
                if flags.y_is_same_or_positive_short() {
                    parser.read::<u8>("Coordinate")?;
                } else {
                    parser.read::<NegativeU8>("Coordinate")?;
                }
            } else {
                if flags.y_is_same_or_positive_short() {
                    // Nothing.
                } else {
                    parser.read::<i16>("Coordinate")?;
                }
            }
        }
        parser.end_group();
    }

    Ok(())
}

fn parse_composite_glyph(parser: &mut Parser) -> Result<()> {
    let flags = parser.read::<CompositeGlyphFlags>("Flag")?;
    parser.read::<GlyphId>("Glyph ID")?;

    let mut matrix = [0.0; 6];

    parser.begin_group("");
    if flags.args_are_xy_values() {
        if flags.arg_1_and_2_are_words() {
            matrix[4] = parser.read::<i16>("E")? as f32;
            matrix[5] = parser.read::<i16>("F")? as f32;
        } else {
            matrix[4] = parser.read::<i8>("E")? as f32;
            matrix[5] = parser.read::<i8>("F")? as f32;
        }
    }

    if flags.we_have_a_two_by_two() {
        matrix[0] = parser.read::<F2DOT14>("A")?.to_f32();
        matrix[1] = parser.read::<F2DOT14>("B")?.to_f32();
        matrix[2] = parser.read::<F2DOT14>("C")?.to_f32();
        matrix[3] = parser.read::<F2DOT14>("D")?.to_f32();
    } else if flags.we_have_an_x_and_y_scale() {
        matrix[0] = parser.read::<F2DOT14>("A")?.to_f32();
        matrix[3] = parser.read::<F2DOT14>("D")?.to_f32();
    } else if flags.we_have_a_scale() {
        matrix[0] = parser.read::<F2DOT14>("A")?.to_f32();
        matrix[3] = matrix[0];
    }

    parser.end_group_with_title(
        format!(
            "Matrix ({} {} {} {} {} {})",
            matrix[0], matrix[1], matrix[2], matrix[3], matrix[4], matrix[5],
        )
    );

    if flags.more_components() {
        parse_composite_glyph(parser)?;
    } else if flags.we_have_instructions() {
        let number_of_instructions = parser.read::<u16>("Number of instructions")?;
        parser.read_bytes(number_of_instructions as usize, "Instructions")?;
    }

    Ok(())
}
